src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -225,9 +249,87 @@ private void pre20Upgrade() throws Exception {
         }
     }
+
+    /**
+     * Loads the current meta state for each index in the new cluster state and checks if it has to be persisted.
+     * Each index state that should be written to disk will be returned. This is only run for data only nodes.
+     * It will return only the states for indices that actually have a shard allocated on the current node.
+     *
+     * @param previouslyWrittenIndices    A list of indices for which the state was already written before
+     * @param potentiallyUnwrittenIndices The list of indices for which state should potentially be written
+     * @param previousMetaData            The last meta data we know of. meta data for all indices in previouslyWrittenIndices list is persisted now
+     * @param newMetaData                 The new metadata
+     * @return iterable over all indices states that should be written to disk
+     */
+    public static Iterable<GatewayMetaState.IndexMetaWriteInfo> filterStates(Set<String> previouslyWrittenIndices, Set<String> potentiallyUnwrittenIndices, MetaData previousMetaData, MetaData newMetaData) {
+        List<GatewayMetaState.IndexMetaWriteInfo> indicesToWrite = new ArrayList<>();
+        for (String index : potentiallyUnwrittenIndices) {
+            IndexMetaData newIndexMetaData = newMetaData.index(index);
+            IndexMetaData previousIndexMetaData = previousMetaData == null ? null : previousMetaData.index(index);
+            String writeReason = null;
+            if (previouslyWrittenIndices.contains(index) == false || previousIndexMetaData == null) {
+                writeReason = "freshly created";
+            } else if (previousIndexMetaData.version() != newIndexMetaData.version()) {
+                writeReason = "version changed from [" + previousIndexMetaData.version() + "] to [" + newIndexMetaData.version() + "]";
+            }
+            if (writeReason != null) {
+                indicesToWrite.add(new GatewayMetaState.IndexMetaWriteInfo(newIndexMetaData, previousIndexMetaData, writeReason));
+            }
+        }
+        return indicesToWrite;
+    }
+
+    public static Set<String> getRelevantIndicesOnDataOnlyNode(ClusterState state, Set<String> previouslyWrittenIndices) {
+        RoutingNode newRoutingNode = state.getRoutingNodes().node(state.nodes().localNodeId());
+        if (newRoutingNode == null) {
+            throw new ElasticsearchIllegalStateException("cluster state does not contain this node - cannot write index meta state");
+        }
+        Set<String> indices = new HashSet<>();
+        for (MutableShardRouting routing : newRoutingNode) {
+            indices.add(routing.index());
+        }
+        // we have to check the meta data also: closed indices will not appear in the routing table, but we must still write the state if we have it written on disk previously
+        for (ObjectCursor<String> index : state.metaData().indices().keys()) {
```

Comment:
we can iterate on the state.metaData() it self. will make this loop slightly simpler